Install library

In [1]:
!pip install pandas

Memanggil library yang digunakan

In [2]:
import pandas as pd

Import data csv/txt/xls dari file directory menggunakan library pandas

In [3]:
data = pd.read_csv('/content/test.csv')

Menampilkan 5 baris teratas bertujuan: agar mendapat penggambaran data apa saja yang kita miliki

In [4]:
data.head()

Id  MSSubClass MSZoning  ...  YrSold  SaleType SaleCondition
0  1461          20       RH  ...    2010        WD        Normal
1  1462          20       RL  ...    2010        WD        Normal
2  1463          60       RL  ...    2010        WD        Normal
3  1464          60       RL  ...    2010        WD        Normal
4  1465         120       RL  ...    2010        WD        Normal

[5 rows x 80 columns]

Melihat apakah ada nilai berupa NA/NaN dalam data

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1455 non-null   object 
 3   LotFrontage    1232 non-null   float64
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          107 non-null    object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1457 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

Dilihat dari RangeIndex seharusnya data entries/baris adalah 1459 non-null (not NA/NaN) tersebar pada 80 kolom. Tetapi pada data.info() terlihat bahwa beberapa kolom tidak memunculkan jumlah entries(jumlah baris) yang sama, sehingga hal ini menunjukkan bahwa pada kolom tersebut ada nilai NA/NaN 

Jika ingin menghapus semua kolom yang mengandung nilai NA/NaN

In [6]:
data_f = data.dropna(axis=1)
#melihat list data_f
data_f

Id  MSSubClass  LotArea Street  ... MiscVal MoSold YrSold SaleCondition
0     1461          20    11622   Pave  ...       0      6   2010        Normal
1     1462          20    14267   Pave  ...   12500      6   2010        Normal
2     1463          60    13830   Pave  ...       0      3   2010        Normal
3     1464          60     9978   Pave  ...       0      6   2010        Normal
4     1465         120     5005   Pave  ...       0      1   2010        Normal
...    ...         ...      ...    ...  ...     ...    ...    ...           ...
1454  2915         160     1936   Pave  ...       0      6   2006        Normal
1455  2916         160     1894   Pave  ...       0      4   2006       Abnorml
1456  2917          20    20000   Pave  ...       0      9   2006       Abnorml
1457  2918          85    10441   Pave  ...     700      7   2006        Normal
1458  2919          60     9627   Pave  ...       0     11   2006        Normal

[1459 rows x 47 columns]

Jika ingin menghapus semua baris yang mengandung nilai NA/NaN

In [7]:
data_p = data.dropna(axis=0)
#melihat list data_p
data_p

Empty DataFrame
Columns: [Id, MSSubClass, MSZoning, LotFrontage, LotArea, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, OverallQual, OverallCond, YearBuilt, YearRemodAdd, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, MasVnrArea, ExterQual, ExterCond, Foundation, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinSF1, BsmtFinType2, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, Heating, HeatingQC, CentralAir, Electrical, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, KitchenQual, TotRmsAbvGrd, Functional, Fireplaces, FireplaceQu, GarageType, GarageYrBlt, GarageFinish, GarageCars, GarageArea, GarageQual, GarageCond, PavedDrive, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, PoolQC, Fence, MiscFeature, MiscVal, MoSold, YrSold, SaleType, SaleCondition]
Index: []

Jika ingin menghapus data NaN pada suatu baris/kolom tertentu

In [8]:
data_s = data.dropna(how='any')
#melihat list data_s
data_s

Empty DataFrame
Columns: [Id, MSSubClass, MSZoning, LotFrontage, LotArea, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, OverallQual, OverallCond, YearBuilt, YearRemodAdd, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, MasVnrArea, ExterQual, ExterCond, Foundation, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinSF1, BsmtFinType2, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, Heating, HeatingQC, CentralAir, Electrical, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, KitchenQual, TotRmsAbvGrd, Functional, Fireplaces, FireplaceQu, GarageType, GarageYrBlt, GarageFinish, GarageCars, GarageArea, GarageQual, GarageCond, PavedDrive, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, PoolQC, Fence, MiscFeature, MiscVal, MoSold, YrSold, SaleType, SaleCondition]
Index: []

Jika ingin menghapus semua baris/kolom yang mengandung NaN

In [9]:
data_d = data.dropna(how='all')
#melihat list data_d
data_d

Id  MSSubClass MSZoning  ...  YrSold  SaleType SaleCondition
0     1461          20       RH  ...    2010        WD        Normal
1     1462          20       RL  ...    2010        WD        Normal
2     1463          60       RL  ...    2010        WD        Normal
3     1464          60       RL  ...    2010        WD        Normal
4     1465         120       RL  ...    2010        WD        Normal
...    ...         ...      ...  ...     ...       ...           ...
1454  2915         160       RM  ...    2006        WD        Normal
1455  2916         160       RM  ...    2006        WD       Abnorml
1456  2917          20       RL  ...    2006        WD       Abnorml
1457  2918          85       RL  ...    2006        WD        Normal
1458  2919          60       RL  ...    2006        WD        Normal

[1459 rows x 80 columns]

Jika ingin mengisi data NaN dengan nilai rata-rata (Apabila list dari data berupa kategori, maka perlu dilakukan konversi variabel kategory (misal a,b,c,d) ke variable indikator (0, 1, 2, dst), namun apabila nilai list pada data keseluruhan adalah angka bukan kategori, maka konversi tidak perlu dilakukan)

In [10]:
#Konversi variabel kategori ke variabel indikator
dataset = pd.get_dummies(data)
# Mengisi data NaN dengan nilai rata-rata:
dataset = dataset.fillna(data.mean())
#melihat list dataset:
dataset

Id  MSSubClass  ...  SaleCondition_Normal  SaleCondition_Partial
0     1461          20  ...                     1                      0
1     1462          20  ...                     1                      0
2     1463          60  ...                     1                      0
3     1464          60  ...                     1                      0
4     1465         120  ...                     1                      0
...    ...         ...  ...                   ...                    ...
1454  2915         160  ...                     1                      0
1455  2916         160  ...                     0                      0
1456  2917          20  ...                     0                      0
1457  2918          85  ...                     1                      0
1458  2919          60  ...                     1                      0

[1459 rows x 271 columns]

Melihat apakah ada nilai berupa NA/NaN dalam data setelah diisi dengan nilai rerata

In [11]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Columns: 271 entries, Id to SaleCondition_Partial
dtypes: float64(11), int64(26), uint8(234)
memory usage: 755.3 KB


Menampilkan jumlah nilai NA/NaN dalam data per kolom

In [12]:
print(dataset.isnull().sum())

Id                       0
MSSubClass               0
LotFrontage              0
LotArea                  0
OverallQual              0
                        ..
SaleCondition_AdjLand    0
SaleCondition_Alloca     0
SaleCondition_Family     0
SaleCondition_Normal     0
SaleCondition_Partial    0
Length: 271, dtype: int64


Jika ingin mengganti nilai NaN dengan data hasil forward propagation, yaitu mengambil acuan dari data urutan diatasnya

In [13]:
#Konversi variabel kategori ke variabel indikator
dataset_2 = pd.get_dummies(data)
# Mengisi data NaN dengan mengacu dari data urutan diatasnya:
dataset_2 = dataset_2.fillna(method='ffill')
#melihat list dataset:
dataset_2

Id  MSSubClass  ...  SaleCondition_Normal  SaleCondition_Partial
0     1461          20  ...                     1                      0
1     1462          20  ...                     1                      0
2     1463          60  ...                     1                      0
3     1464          60  ...                     1                      0
4     1465         120  ...                     1                      0
...    ...         ...  ...                   ...                    ...
1454  2915         160  ...                     1                      0
1455  2916         160  ...                     0                      0
1456  2917          20  ...                     0                      0
1457  2918          85  ...                     1                      0
1458  2919          60  ...                     1                      0

[1459 rows x 271 columns]

Jika ingin mengganti nilai NaN dengan data hasil back propagation, yaitu mengambil acuan dari data urutan dibawahnya

In [14]:
#Konversi variabel kategori ke variabel indikator
dataset_3 = pd.get_dummies(data)
# Mengisi data NaN dengan mengacu dari data urutan dibawahnya:
dataset_3 = dataset_3.fillna(method='bfill')
#melihat list dataset:
dataset_3

Id  MSSubClass  ...  SaleCondition_Normal  SaleCondition_Partial
0     1461          20  ...                     1                      0
1     1462          20  ...                     1                      0
2     1463          60  ...                     1                      0
3     1464          60  ...                     1                      0
4     1465         120  ...                     1                      0
...    ...         ...  ...                   ...                    ...
1454  2915         160  ...                     1                      0
1455  2916         160  ...                     0                      0
1456  2917          20  ...                     0                      0
1457  2918          85  ...                     1                      0
1458  2919          60  ...                     1                      0

[1459 rows x 271 columns]

Jika ingin mengganti nilai NaN dengan angka 0

In [15]:
#Konversi variabel kategori ke variabel indikator
dataset_4 = pd.get_dummies(data)
# Mengisi data NaN dengan mengacu dari data urutan dibawahnya:
dataset_4 = dataset_4.fillna(0)
#melihat list dataset:
dataset_4

Id  MSSubClass  ...  SaleCondition_Normal  SaleCondition_Partial
0     1461          20  ...                     1                      0
1     1462          20  ...                     1                      0
2     1463          60  ...                     1                      0
3     1464          60  ...                     1                      0
4     1465         120  ...                     1                      0
...    ...         ...  ...                   ...                    ...
1454  2915         160  ...                     1                      0
1455  2916         160  ...                     0                      0
1456  2917          20  ...                     0                      0
1457  2918          85  ...                     1                      0
1458  2919          60  ...                     1                      0

[1459 rows x 271 columns]

Jika ingin menggunakan interpolasi linear (dengan menggunakan konversi kategori ke indikator)

In [16]:
#Konversi variabel kategori ke variabel indikator
dataset_5 = pd.get_dummies(data)
# Metode interpolasi linear
dataset_5 = dataset_5.interpolate(method='linear')
#melihat list dataset:
dataset_5

Id  MSSubClass  ...  SaleCondition_Normal  SaleCondition_Partial
0     1461          20  ...                     1                      0
1     1462          20  ...                     1                      0
2     1463          60  ...                     1                      0
3     1464          60  ...                     1                      0
4     1465         120  ...                     1                      0
...    ...         ...  ...                   ...                    ...
1454  2915         160  ...                     1                      0
1455  2916         160  ...                     0                      0
1456  2917          20  ...                     0                      0
1457  2918          85  ...                     1                      0
1458  2919          60  ...                     1                      0

[1459 rows x 271 columns]

Jika ingin menggunakan interpolasi linear (tanpa menggunakan konversi kategori ke indikator) = disini nilai NaN masih ada, karena interpolasi bukan mengganti/mengisi data NaN melainkan mengganti data tertentu sesuai hasil interpolasi linear.

In [20]:
dataset_6 = data.interpolate(method='linear')
#melihat list dataset:
dataset_6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1455 non-null   object 
 3   LotFrontage    1459 non-null   float64
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          107 non-null    object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1457 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

Solusi masalah diatas yaitu dengan menggunakan nilai threshold pada interpolasi metodi pad untuk membatasi maksimum jumlah NaN dikehendaki.



In [19]:
dataset_7 = data.interpolate(method='pad', limit=7)
#melihat list dataset:
dataset_7.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1455 non-null   object 
 3   LotFrontage    1232 non-null   float64
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          1459 non-null   object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1459 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

NaN sudah berkurang namun masih terdapat NaN pada 3 kolom, dimana jumlah baris yang non Nan bukan 1459:
1.   LotFrontage    1232 non-null   float64 
2.   MasVnrArea     1455 non-null   float64
3.   MSZoning       1455 non-null   object







Karena kedua kolom tersebut berisi nilai float atau angka dan satu kolom berisi object, sehingga solusi diatas yaitu kembali ke solusi sebelumnya yaitu dengan mengisi nilai angka dengan rerata, forward propagation, backward propagation, nol, dan object dengan metode interpolasi.

In [29]:
dataset_7['LotFrontage'] = dataset_7['LotFrontage'].fillna(dataset_7.mean)
dataset_7['MasVnrArea'] = dataset_7['MasVnrArea'].fillna(dataset_7.mean)
dataset_7['MSZoning'] = dataset_7['MSZoning'].interpolate(method='pad', limit = 3)
dataset_7.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1459 non-null   object 
 3   LotFrontage    1459 non-null   object 
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          1459 non-null   object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1459 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

In [30]:
print(dataset_7.isnull().sum())

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MiscVal          0
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
Length: 80, dtype: int64


Sehingga kita dapatkan data tanpa NaN 

Selanjutnya, apabila terdapat data yang duplikat, maka cara menghapusnya yaitu

In [24]:
dataset_8 = data.drop_duplicates() # drop duplicate 

Jika ingin mengganti nilai pada data di kolom tertentu

In [22]:
data['MasVnrArea'] = data['MasVnrArea'].replace(['0'], '20')

Jika ingin mengganti nilai pada data keseluruhan

In [23]:
dataset_9 = data.replace(['0'], '20')

Export data hasil cleaning process ke dalam data csv atau excel

In [21]:
dataset_7.to_excel("output.xlsx")



---

# SELESAI!! wohooooooo.....